### PR TITLE
Implement recursive descent parser

### PR DESCRIPTION
**Pull request by mkaminski1988**
@crocodele  This PR replaces the kludgy postfix expression parser with a recursive descent parser, with a net decrease of 132 LoC. :)

The parser is based on what I learned by reading the first few chapters of [Crafting Interpreters](http://craftinginterpreters.com/).

We can now provide users a more formal description of the JSONpath filter expression language supported by the extension.

```
json_path  → TBD ;
child_path → TBD ;

expression → logic_or ;
logic_or   → logic_and ( "||" logic_and )* ;
logic_and  → equality ( "&&" equality )* ;
equality   → comparison ( ( "!=" | "==" ) comparison )* ;
comparison → primary ( ( ">" | ">=" | "<" | "<=", "=~" ) primary )* ;
primary    → NUMBER | STRING | "true" | "false" | "(" expression ")" | json_path | child_path ;
```

After merging, I will work on moving the entire JSONPath parsing logic into the recursive descent parser.

### CODE REVIEW

**Comment by mkaminski1988** [comment]
Whoops, found some memory leaks that I need to close.

**Comment by mkaminski1988** [comment]
Memory leaks fixed.

File: src/jsonpath/parser.c [hunk 1]
```diff
@@ -6,297 +6,69 @@
...
+#define PARSER_PARAMS \
+  lex_token lex_token[PARSE_BUF_LEN], char lex_tok_values[][PARSE_BUF_LEN], int *lex_idx, int lex_tok_count
+
+static struct ast_node* ast_alloc_binary(enum ast_type type, struct ast_node* left, struct ast_node* right);
+static struct ast_node* ast_alloc_node(struct ast_node* prev, enum ast_type type);
+
+static struct ast_node* parse_expression(PARSER_PARAMS);
+static struct ast_node* parse_and(PARSER_PARAMS);
+static struct ast_node* parse_comparison(PARSER_PARAMS);
+static struct ast_node* parse_comparison(PARSER_PARAMS);
```
Comment (crocodele) [inline src/jsonpath/parser.c:26]: Duplicate row.

**Review by crocodele — APPROVED** [review]
Looks nice!

**Comment by crocodele** [comment]
I made an attempt to outlaw multiplication in https://github.com/supermetrics/php-ext-jsonpath/commit/26f45172fac5254155174b92dd363eb77d4f5b33, and noticed that the nodes allocated in the parsing aren't freed if the parsing fails. At least not in this particular case. I suppose it's due to the parser returning `NULL` and the nodes ending up as orphans?

**Comment by mkaminski1988** [comment]
@crocodele The error _should_ propagate back to the caller in `jsonpath.c`, where the call to `free_ast_nodes` is made. Is this not the case?

**Comment by mkaminski1988** [comment]
Ah I see what's happening. Perhaps try `free_ast_nodes(head)`?

**Comment by mkaminski1988** [comment]
Ah I see what's happening. Perhaps try `free_ast_nodes(ret)`?

File: src/jsonpath/parser.c (hunk 1 above)
Comment (mkaminski1988) [reply src/jsonpath/parser.c]: Good catch! Fixed.

**Pull request merged**